File: src/days/Day09.kt
package days

import helpers.FileUtils
import java.awt.Point

object Day09 {
    @JvmStatic
    fun main(args: Array<String>) {
        val raw = rawData()
        val lines = raw.split("\n")
        println("Problem 1: " + problem1(lines))
        println("Problem 2: " + problem2())
    }

    private fun problem1(lines: List<String>): Int {
        val visitedMap = HashSet<Point>()
        val curHeadPos = Point(0, 0)
        val curTailPos = Point(0, 0)
        for (line in lines) {
            val split = line.split(" ")
            val direction: Point = parseDirection(split[0])
            val distance: Int = Integer.parseInt(split[1])

            for (i in 1 .. distance) {
                curHeadPos.translate(direction.x, direction.y)
                if (curHeadPos.distance(curTailPos) >= 2.0) { curTailPos.setLocation(
                    curHeadPos.x-direction.x, curHeadPos.y-direction.y
                ); }
                visitedMap.add(curTailPos)
            }
        }
        return visitedMap.size
    }

    private fun problem2(): Int {
        return 0
    }

    private fun parseDirection(letter: String): Point {
        return when (letter) {
            "U" -> { Point(0, 1) }
            "D" -> { Point(0, 1) }
            "R" -> { Point(1, 0) }
            "L" -> { Point(-1, 0 ) }
            else -> { Point(0, 0) }
        }
    }

    private fun rawData(): String {
        return FileUtils.fileToString("src/inputs/Day09Input.txt")
    }
}

File: src/days/Day09.rs
use std::{collections::HashSet, fs};

const DAY_HEADER: &str = "
██████╗  █████╗ ██╗   ██╗     ██████╗  █████╗ 
██╔══██╗██╔══██╗╚██╗ ██╔╝    ██╔═████╗██╔══██╗
██║  ██║███████║ ╚████╔╝     ██║██╔██║╚██████║
██║  ██║██╔══██║  ╚██╔╝      ████╔╝██║ ╚═══██║
██████╔╝██║  ██║   ██║       ╚██████╔╝ █████╔╝
╚═════╝ ╚═╝  ╚═╝   ╚═╝        ╚═════╝  ╚════╝ 
";

pub fn main() {
    println!("{}", DAY_HEADER);
    let raw = fs::read_to_string("/home/focus/code/AdventOfCode2022/inputs/9").expect("File does not exist");
    problem1(&raw);
    problem2(&raw);
}

fn problem1(raw: &str) {
    let mut visited_map = HashSet::new();
    let mut head = Point::new(0, 0);
    let mut tail = Point::new(0, 0);
    raw.lines().for_each(|line| {
        let split = line.split(" ").collect::<Vec<&str>>();
        let transform: Point = parseDirection(split[0]);
        let distance = split[1].parse::<i32>().unwrap(); 

        (1..=distance).for_each(|_| {
            head.transform(transform.x, transform.y);
            if head.distance(&tail) >= 1.9 {
                tail.transform(transform.x, transform.y);
            }
            let hash = format!("{}:{}", tail.x, tail.y);
            visited_map.insert(hash);
        });
    });
    println!("Problem 1: {}", visited_map.len());
}

fn problem2(raw: &str) {
    todo!();
}

fn parseDirection(letter: &str) -> Point {
    match letter {
        "U" => { Point::new(0, 1) }
        "D" => { Point::new(0, 1) }
        "R" => { Point::new(1, 0) }
        "L" => { Point::new(-1, 0 ) }
        _ => { panic!("Invalid direction") }
    }
}

#[derive(Hash, Eq, PartialEq)]
struct Point {
    x: i32,
    y: i32
}

impl Point {
    fn new(x: i32, y: i32) -> Point {
        Point { x, y }
    }

    fn transform(&mut self, x: i32, y: i32) {
        self.x += x;
        self.y += y;
    }

    fn distance(&self, other: &Point) -> f32 {
        let x = (self.x - other.x).pow(2);
        let y = (self.y - other.y).pow(2);
        ((x + y) as f32).sqrt()
    }
}
